### ISSUE TITLE
`service update` for a 1-replica service leads to poor spreading

### ISSUE THREAD

**Issue by a-robinson** [docker/docker#26826]
**Description**

When updating the arguments to a swarm mode service, the ordering of bringing the old container down and the new container up can lead to poor scheduling decisions in some circumstances.

The general scenario where we've been hitting this is that we create a 3 node swarm mode cluster, then create 3 replicas of cockroachdb in the cluster by making 3 services with one replica each. Up to this point, everything is great -- the 3 replicas get spread across the 3 nodes. Due to the mechanics of cockroachdb cluster joining, we then want to change the command-line arguments to the first service so that it can properly rejoin the cluster if it ever gets moved to a different node, so we run `docker service update` on it. This causes it to get recreated, but this recreation often leaves us with a poorly balanced cluster where 2 of our cockroachdb instances are on the same node and one of the nodes is empty.

It looks like this happens because the new replica gets scheduled and started before the old one is shut down without accounting for the fact that the old one is being shut down, and so all 3 nodes look the same to the scheduler.

**Steps to reproduce the issue:**

You could presumably strip out most of the details of these commands, but for the sake of not removing anything important, this step of commands on a 3-node cluster reliably hits the problem.

1. docker network create --driver overlay cockroachdb
2. docker service create --replicas 1 --name cockroachdb-0 --network cockroachdb --label app=cockroachdb --mount type=volume,source=cockroachdb-0,target=/cockroach/cockroach-data,volume-driver=local --stop-grace-period 60s cockroachdb/cockroach:latest start --logtostderr --insecure
3. docker service create --replicas 1 --name cockroachdb-1 --network cockroachdb --label app=cockroachdb --mount type=volume,source=cockroachdb-1,target=/cockroach/cockroach-data,volume-driver=local --stop-grace-period 60s cockroachdb/cockroach:latest start --logtostderr --insecure --join cockroachdb-0:26257
4. docker service create --replicas 1 --name cockroachdb-2 --network cockroachdb --label app=cockroachdb --mount type=volume,source=cockroachdb-2,target=/cockroach/cockroach-data,volume-driver=local --stop-grace-period 60s cockroachdb/cockroach:latest start --logtostderr --insecure --join cockroachdb-0:26257
5. docker service update cockroachdb-0 --args "start --logtostderr --insecure --join cockroachdb-1:26257"

**Describe the results you received:**

After step 3, the 3 cockroachdb containers are reliably spread across the 3 nodes. After step 4, the cockroachdb-0 container is often on the same node as one of the other two containers, leaving a node empty.

**Describe the results you expected:**

I'd expect the re-created cockroachdb-0 container to end up on the same node that it was on before the update.

**Additional information you deem important (e.g. issue happens only occasionally):**

This is unlikely to repro 100% of the time, probably just 67% of the time. I haven't had any trouble getting it to repro.

**Output of `docker version`:**

Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64

**Output of `docker info`:**

Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:02:53 2016
 OS/Arch:      linux/amd64
root@a1:/home/alex# docker info
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 7
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 37
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: active
 NodeID: 7c8dng9wocp2ahyvdl0ijyjj7
 Is Manager: true
 ClusterID: 5968pkjl89rp2nc9x3376c1dl
 Managers: 1
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.128.0.3
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.625 GiB
Name: a1
ID: UVKD:BC5Y:WI2F:CZY6:4KMQ:E5R6:TP4U:NIMG:254U:FZSV:AWTE:QSPX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
Insecure Registries:
 127.0.0.0/8

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This is on GCE, although I doubt that matters for this issue.

**Comment by a-robinson** [comment]
cc @jseldess

**Comment by aluzzardi** [comment]
Thanks for the report @a-robinson.

We have a new default "HA" strategy coming up in 1.13 which, in addition to the current spreading model, will also spread based on number of tasks from the same service (so you'll never have multiple instances of the same service on one node if there are other nodes available with no instance).

However, since we have different services here it won't do the trick.

@aaronlehmann I think we discussed this issue in the past (e.g. looking at the desired state rather than actual state when performing the scheduling - can't recall the exact issue?).

@a-robinson In the meantime, you could play with constraints (which are not an ideal solution for you). 

/cc @aaronlehmann @dperny @dongluochen

**Comment by aaronlehmann** [comment]
> @aaronlehmann I think we discussed this issue in the past (e.g. looking at the desired state rather than actual state when performing the scheduling - can't recall the exact issue?).

This is https://github.com/docker/swarmkit/pull/874

The HA scheduler in 1.13 looks at desired state rather than actual state when making a scheduling decision. So this should be fixed in 1.13. @a-robinson, would you be interested in testing with a nighly build of Docker 1.13 from https://master.dockerproject.org to see if the scheduling behavior has improved?

**Comment by a-robinson** [comment]
Thanks @aluzzardi @aaronlehmann. I'll test it out with a recent build when I have a little more time later this week.

**Comment by a-robinson** [comment]
The Sept 30 build is exhibiting the ideal behavior for us. I can repeatedly change the arguments of a 1-replica service and it's getting recreated on the same VM every time. Thanks!

**Comment by thaJeztah** [comment]
Thanks @a-robinson, thats good to hear
